### PR TITLE
.github: add and use ccache on cygwin build

### PR DESCRIPTION
**Pull request by khancyr**
Enable ccache on cygwin
Half cygwin CI time !

### CODE REVIEW

**Comment by magicrub** [comment]
I'm tempted to merge this but I don't have any means of verifying it other than seeing that it ran successfully on CI. Is that good enough? Probably...???

 PR https://github.com/ArduPilot/ardupilot/pull/22932 was just merged so this should get rebased to make sure it still passes CI

**Comment by magicrub** [comment]
@khancyr I rebased for you.

**Comment by magicrub** [comment]
Any objections about merging this? I'm not 100% confident to do it myself.

File: .github/workflows/cygwin_build.yml [hunk 1]
```diff
@@ -13,21 +13,65 @@ jobs:
...
+          packages: cygwin64 gcc-g++=10.2.0-1 ccache python37 python37-future python37-lxml python37-pip python37-setuptools python37-wheel git procps gettext
+          add-to-path: false
+      # Put ccache into github cache for faster build
+      - name: setup ccache
+        env:
+          PATH: /usr/bin:$(cygpath ${SYSTEMROOT})/system32
+        shell: C:\cygwin\bin\bash.exe -eo pipefail '{0}'
+        run: >-
+          mkdir -p /cygdrive/d/a/ardupilot/ardupilot/ccache &&
+          mkdir -p /usr/local/etc &&
```
Comment (peterbarker) [inline .github/workflows/cygwin_build.yml:37]: What's this for?

File: .github/workflows/cygwin_build.yml [hunk 2]
```diff
@@ -13,21 +13,65 @@ jobs:
...
-          packages: cygwin64 gcc-g++=10.2.0-1 python37 python37-future python37-lxml python37-pip python37-setuptools python37-wheel git procps gettext
+          packages: cygwin64 gcc-g++=10.2.0-1 ccache python37 python37-future python37-lxml python37-pip python37-setuptools python37-wheel git procps gettext
+          add-to-path: false
+      # Put ccache into github cache for faster build
+      - name: setup ccache
+        env:
+          PATH: /usr/bin:$(cygpath ${SYSTEMROOT})/system32
+        shell: C:\cygwin\bin\bash.exe -eo pipefail '{0}'
+        run: >-
+          mkdir -p /cygdrive/d/a/ardupilot/ardupilot/ccache &&
```
Comment (peterbarker) [inline .github/workflows/cygwin_build.yml:36]: Why all of the `&&` through here rather than `set -ex` at the top?

Similarly for the other blocks.

**Review by peterbarker — CHANGES_REQUESTED** [review]
We definitely want this if it halves the cygwin build times!

File: .github/workflows/cygwin_build.yml (hunk 2 above)
Comment (khancyr) [reply .github/workflows/cygwin_build.yml:36]: For the && , that is the recommanded way to do multiline into the cygwin subshell. https://github.com/cygwin/cygwin-install-action#line-endings

File: .github/workflows/cygwin_build.yml (hunk 1 above)
Comment (khancyr) [reply .github/workflows/cygwin_build.yml:37]: leftover... should be /usr/local/bin/ now

**Comment by khancyr** [comment]
unfortunatelly, it doesn't half the whole CI. There is some gains but not as much as expected ... which is weird as ccache is working well if I test it on a local cygwin machine

**Comment by peterbarker** [comment]
> unfortunatelly, it doesn't half the whole CI. There is some gains but not as much as expected ... which is weird as ccache is working well if I test it on a local cygwin machine

So it seems to save ~8m in this PR.

We really need timings here.  IIRC one of the problems we've previously had with ccache is that the time spent storing the cache outweighed the time saved from it.  It's currently only 56MB - perhaps the same thing will apply later.

Is adding timing hard?  Even just timestamps are good..

**Comment by khancyr** [comment]
cache isn't an issue on github action runner : 
image
7s for 600MB on restaure
32s to upload the cache
so against a few minutes we still win

ccache cache is caped to 400MB so it won't be worst than this

**Comment by peterbarker** [comment]
@khancyr thanks for the timing information.  Merged this, let's see how it goes!

**Pull request merged**